src/gf.c

```diff
@@ -2989,35 +2639,303 @@ static jl_value_t *ml_matches(jl_methtable_t *mt, int offs,
             return env.t;
         }
     }
-    htable_new(&env.visited, 0);
-    if (((jl_datatype_t*)unw)->isdispatchtuple) {
-        search.min_valid = env.min_valid;
-        search.max_valid = env.max_valid;
-        jl_typemap_entry_t *ml = jl_typemap_assoc_by_type(defs, &search, offs, /*subtype*/1);
-        env.min_valid = search.min_valid;
-        env.max_valid = search.max_valid;
-        if (ml) {
-            env.match.ti = (jl_value_t*)type;
-            env.match.env = search.env;
-            env.match.issubty = 1;
-            env.match.fptr(ml, &env.match); // also matches all ambig and prior, as needed
+    jl_typemap_intersection_visitor(defs, offs, &env.match);
```

Comment:
It would be good to figure out some sort of limit on this --- sometimes it returns a large number of items during inference, and we're unlikely to cut it down to 3. For example `==(::Any, ::Any)`. Maybe a larger arbitrary cutoff? Or maybe count how many methods had dispatchtuple signatures, since we know those are all disjoint, and early-out based on that?

Reply:
Most efficient would perhaps be to change `lim` just to mean "number of intersecting signatures". I was slightly surprised I managed to preserve (and refine) the original meaning here. There's an early-exit in the first `O(n^2)` loop, which helps considerably for that case. Counting `dispatchtuple` signatures seems like a simple addition to that.

```
julia> @btime Base._methods_by_ftype(Tuple{typeof(==), Any, Any}, -1, Base.get_world_counter())
  76.585 μs (194 allocations: 12.88 KiB) # master
  17.504 ms (46063 allocations: 2.56 MiB) # now (258d04e8bfdb8189fd61199bccd7641650fec07b)
  17.386 ms (46063 allocations: 2.56 MiB) # with earlier limit from counting dispatch-tupletypes matches (ec8961c77614b80d21ee51d6353769c9ae57d899)

julia> @btime Base._methods_by_ftype(Tuple{typeof(==), Any, Any}, 4, Base.get_world_counter())
  3.824 μs (8 allocations: 528 bytes) # master
  280.331 μs (858 allocations: 76.42 KiB) # PR
  880.929 ns (9 allocations: 624 bytes) # with limit
```